internal/pkg/gateway/registry/registry.go

```diff
@@ -7,40 +7,218 @@ SPDX-License-Identifier: Apache-2.0
-func (reg *Registry) Orderers(channel string) []orderer.AtomicBroadcast_BroadcastClient {
-	// placeholder code for now - just return empty array - Submit() won't work at the moment.
-	// this will be replaced by the set of orderers for a channel determined from the discovery service
-	return make([]orderer.AtomicBroadcast_BroadcastClient, 0)
+func (reg *Registry) Orderers(channel string) ([]endpoint.Orderer, error) {
+	reg.registerChannel(channel)
+	var orderers []endpoint.Orderer
+
+	// Get the config
+	config, err := reg.discovery.Config(channel)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, eps := range config.GetOrderers() {
+		for _, ep := range eps.Endpoint {
+			url := fmt.Sprintf("%s:%d", ep.Host, ep.Port)
+			if ord, ok := reg.orderers[url]; ok {
+				orderers = append(orderers, ord)
+			}
+		}
+	}
+
+	return orderers, nil
+}
+
+func (reg *Registry) addPeer(mspid string, url string) error {
+	// assumes that the msp registry has already been populated with this mspid
+	if _, ok := reg.peers[url]; !ok && strings.Contains(url, ":") {
+		// this peer is new - connect to it and add to the peers registry
+		endorser, err := reg.newRemoteEndorser(url, mspid)
+		if err != nil {
+			return err
+		}
+		reg.peers[url] = endorser
+		reg.logger.Infof("Added peer to registry: %s", url)
+	}
+
+	return nil
+}
+
+func (reg *Registry) addOrderer(mspid string, host string, port uint32) error {
+	// assumes that the msp registry has already been populated with this mspid
+	url := fmt.Sprintf("%s:%d", host, port)
+	if _, ok := reg.orderers[url]; !ok {
+		// this orderer is new - connect to it and add to the orderers registry
+		orderer, err := reg.newOrderer(url, mspid)
+		if err != nil {
+			return err
+		}
+		reg.orderers[url] = orderer
+		reg.logger.Infof("Added orderer to registry: %s", url)
+	}
+	return nil
+}
+
+func (reg *Registry) registerChannel(channel string) error {
+	// todo need to handle membership updates
+	init, ok := reg.channelsInitialized[channel]
+	if !init || !ok {
+		// get config and peer discovery info for this channel
+
+		// Get the config
+		config, err := reg.discovery.Config(channel)
+		if err != nil {
+			return err
+		}
+		// get the tlscerts
+		for msp, info := range config.GetMsps() {
+			reg.msps[msp] = info.GetTlsRootCerts()[0] // todo handle more than one
+		}
+
+		for msp, eps := range config.GetOrderers() {
+			for _, ep := range eps.Endpoint {
+				err = reg.addOrderer(msp, ep.Host, ep.Port)
+				if err != nil {
+					reg.logger.Errorf("Error adding orderer to registry: %s", err)
+				}
+			}
+		}
+
+		// get the peers for the channel
+		peers := map[string]string{}
+		members := reg.discovery.PeersOfChannel(common.ChannelID(channel))
+		for _, member := range members {
+			id := member.PKIid.String()
+			peers[id] = member.PreferredEndpoint()
+		}
+		for org, infoset := range reg.discovery.IdentityInfo().ByOrg() {
+			for _, info := range infoset {
+				pkid := info.PKIId.String()
```

Comment:
This is a landmine. We're hoping that the `fmt.Stringer` implementation on a byte slice that is documented as
```
// PKIidType defines the type that holds the PKI-id
// which is the security identifier of a peer
```
(ie. opaque) is going to be an address? And we're assuming that it's going to stay that way?

Reply:
It's only using the pkid to correlate the peers into groups of the same mspid.  It makes no assumption as to whether it is an address or not.  I copied this code from the discovery service which does something similar.

Reply:
A PKI-ID is not an address. It's a hash of the serialized identity.